ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,321 @@
+        base_model,
+        torch_dtype=torch.float16,
+        device_map={"": "cpu"},
+    )
+```
+
+> **Note**
+>
+> In the merging state, load_in_low_bit="nf4" should be removed since we need to load the original model as the base model.
+
+
+
+### 7.1.3.2 Merge the Weights
+
+
+
+Then we can load the QLoRA weights to enable the merging process.
+
+```python
+from bigdl.llm.transformers.qlora import PeftModel
+adapter_path = "./outputs/checkpoint-200"
+lora_model = PeftModel.from_pretrained(
+        base_model,
+        adapter_path,
+        device_map={"": "cpu"},
+        torch_dtype=torch.float16,
+    )
+
+
+```
+> **Note**
+>
+> Instead of from peft import PeftModel, we import PeftModel from bigdl.llm.transformers.qlora as a BigDL-LLM compatible model. 
+> 
+> **Note**
+> The adapter path is the local path you save the fine-tuned model, in our case is `./outputs/checkpoint-200`.
+>
+
+To verify if the LoRA weights have worked in conjunction with the pretrained model, the first layer weights (which in llama2 case are trainable queries) are extracted to highlight the difference.
+
+```python
+first_weight = base_model.model.layers[0].self_attn.q_proj.weight
+first_weight_old = first_weight.clone()
+lora_weight = lora_model.base_model.model.model.layers[0].self_attn.q_proj.weight
+assert torch.allclose(first_weight_old, first_weight)
+```
+With the new merging method `merge_and_unload`, we can easily combine the fine-tuned model with pre-trained model, and testify whether the weights have changed with the `assert` statement. 
+
+```python
+lora_model = lora_model.merge_and_unload()
+lora_model.train(False)
+assert not torch.allclose(first_weight_old, first_weight)
+```
+You may get the outputs below without error report to indicate the successful conversion.
+```
+Using pad_token, but it is not set yet.
+Using pad_token, but it is not set yet.
+```
+Finally we can save the fine-tuned model in a specified local path (in our case is `./outputs/checkpoint-200-merged`).
+```python
+output_path = ./outputs/checkpoint-200-merged
+lora_model_sd = lora_model.state_dict()
+deloreanized_sd = {
+        k.replace("base_model.model.", ""): v
+        for k, v in lora_model_sd.items()
+        if "lora" not in k
+    }
+base_model.save_pretrained(output_path, state_dict=deloreanized_sd)
+tokenizer.save_pretrained(output_path)
+
+```
+
+
+## 7.1.4 Inference with Fine-tuned model
+After merging and deploying the models, we can test the performance of the fine-tuned model. 
+The detailed instructions of running
+LLM inference with BigDL-LLM optimizations could be found in [Chapter 6](../ch_6_GPU_Acceleration/6_1_GPU_Llama2-7B.md), here we quickly go through the preparation of model inference.
+
+### 7.1.4.1 Inference with the Fine-tuned Model
+```python
+model_path = "./outputs/checkpoint-200-merged"
+model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path = model_path,load_in_4bit=True)
+model.config.use_cache = True
```

Comment:
Do we need explicitly emphasize here `use_cache=True`?

Reply:
According to [the example](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py), this line can be removed here